_drafts/cortex-m-breakpoints.md

```diff
@@ -0,0 +1,607 @@
+(gdb) c
+Continuing.
+
+[...]
+
+Breakpoint 8, dummy_function_7 () at dummy_functions.c:28
+28    EXAMPLE_LOG("stub function '%s' called", __func__);
+(gdb) c
+Continuing.
+
+Breakpoint 2, dummy_function_ram () at dummy_functions.c:41
+41    EXAMPLE_LOG("stub function '%s' called", __func__);
+(gdb) c
+```
+
+Neat, they all worked so a breakpoint was installed for our flash-based function, `dummy_function_1`, that was missing from the FPB configuration.
+
+Let's dump the functions to examine if any instructions were modified:
+
+```
+shell> dump_dummy_funcs
+dummy_function_1: Starts at 0x240. First Instruction = 0x4802be00
+dummy_function_2: Starts at 0x250. First Instruction = 0x48024901
+dummy_function_3: Starts at 0x260. First Instruction = 0x48024901
+dummy_function_4: Starts at 0x270. First Instruction = 0x48024901
+dummy_function_5: Starts at 0x280. First Instruction = 0x48024901
+dummy_function_6: Starts at 0x290. First Instruction = 0x48024901
+dummy_function_7: Starts at 0x2a0. First Instruction = 0x48024901
+dummy_function_8: Starts at 0x2b0. First Instruction = 0x48024901
+dummy_function_9: Starts at 0x2c0. First Instruction = 0x48024901
+dummy_function_ram: Starts at 0x20000000. First Instruction = 0x4802be00
+```
+
+Aha, we can see `dummy_function_1` also was updated to include a breakpoint instruction, `0xbe00`. What's interesting here is this address is in the NRF52's internal flash space. This means to inject this breakpoint the flash page the address is in must have been reprogrammed.
+
+This is in fact exactly what happened. The SEGGER JLinkGDBServer implements a feature referred to in the docs as "unlimited flash breakpoints"[^11] which will patch flash regions the J-Link Software knows how to write to with `bkpt` instructions.
+
+#### Flash Software Breakpoint Optimizations
+
+To minimize the amount of times the flash sector has to be rewritten, the JLinkGDBServer will also perform some additional nifty optimizations such as caching the flash sector and not actually uninstalling the breakpoint when gdb requests. When a request to single-step over the breakpoint is actually made, the gdbserver will "simulate" the instruction execution. For example if we objdump our dummy function, we can see the instruction patched over, `0x4901`, is `ldr r1, [pc, #4]`:
+
+```bash
+$ arm-none-eabi-objdump  --disassemble=dummy_function_1  \
+    example/breakpoint/build/nrf52.elf
+
+build/nrf52.elf:     file format elf32-littlearm
+
+
+Disassembly of section .text:
+
+00000240 <dummy_function_1>:
+ 240:	4901        ldr	r1, [pc, #4] ; (248 <dummy_function_1+0x8>)
+ 242:	4802        ldr	r0, [pc, #8] ; (24c <dummy_function_1+0xc>)
+ 244:	f000 b966   b.w	514 <example_log>
+ 248:	0000732f    .word	0x0000732f
+ 24c:	00007315    .word	0x00007315
+```
+
+From a debugger we could perform the `pc`-relative load into `r1` and increment the `pc` instead of re-flashing the sector with the real instruction so the hardware can execute it.
```

Comment:
As mentioned, this is not clear at all. You should expand on it as it is super cool!

Reply:
tried to be more descriptive, lmk what you think!